src/sparse/linalg/ordering.rs

```diff
@@ -13,74 +13,223 @@ pub struct Ordering<I> {
-                min_deg_vert = vert;
+        // Write the next permutation in reverse order.
+        perm[nb_vertices - perm_index - 1] = I::from_usize(current_vertex);
+        visited[current_vertex.index()] = true;
+
+        // Find, sort, and push all new neighbors of the current vertex.
+        let outer = mat.outer_view(current_vertex.index()).unwrap();
+        neighbors.clear();
+        for &neighbor in outer.indices() {
+            if !visited[neighbor.index()] {
+                neighbors.push((degrees[neighbor.index()], neighbor));
+                visited[neighbor.index()] = true;
             }
         }
-        deque.clear();
-        deque.push_back(min_deg_vert);
 
-        while let Some(cur_vert) = deque.pop_front() {
-            if visited[cur_vert] {
-                continue;
-            }
-            perm.push(I::from_usize(cur_vert));
-            visited[cur_vert.index()] = true;
-            let outer = mat.outer_view(cur_vert.index()).unwrap();
-            neighbors.clear();
-            for &neighbor in outer.indices() {
-                if !visited[neighbor.index()] {
-                    neighbors.push((degrees[neighbor.index()], neighbor));
-                }
-            }
-            neighbors.sort_by_key(|&(deg, _)| deg);
-            for (_deg, neighbor) in &neighbors {
-                deque.push_back(neighbor.index());
-            }
+        // Missed optimization: match small sizes explicitly, sort using sorting networks.
+        // This especially makes sense if swaps are predictably compiled into cmov instructions,
+        // which they aren't currently, see https://github.com/rust-lang/rust/issues/53823.
+        // For more information on how to do sorting networks efficiently see https://arxiv.org/pdf/1505.01962.pdf.
+        neighbors.sort_unstable_by_key(|&(deg, _)| deg);
+
+        for (_deg, neighbor) in &neighbors {
+            deque.push_back(neighbor.index());
         }
-        connected_parts.push(perm.len());
     }
 
+    connected_parts.push(nb_vertices);
+
     Ordering {
         perm: PermOwnedI::new(perm),
-        connected_parts,
+        connected_parts: connected_parts
+            .into_iter()
+            .map(|i| nb_vertices - i)
+            .rev()
+            .collect(),
+    }
+}
+
+fn find_pseudoperipheral_vertex<N, I, Iptr>(
+    visited: &[bool],
+    degrees: &[usize],
+    mat: &CsMatViewI<N, I, Iptr>,
+) -> Option<usize>
+where
+    N: PartialEq,
+    I: SpIndex,
+    Iptr: SpIndex,
+{
+    // Choose the next available vertex as currrent starting vertex.
+    let mut current = visited
+        .iter()
+        .enumerate()
+        .find(|(_i, &a)| !a)
+        .map(|(i, _a)| i)?;
+
+    // Isolated vertices are by definition pseudoperipheral.
+    if degrees[current] == 0 {
+        return Some(current);
+    }
+
+    let (mut contender, mut current_height) =
+        rls_contender_and_height(current, degrees, mat);
+
+    loop {
```

Comment:
Could you add comments explaining why this loop should terminate?

Reply:
A convincing, more or less formal argument is that, for the loop to continue spinning, the height of the rls of the currently selected vertex must always strictly increase. However, in any connected commponent there is always a set of vertices whose rls's height is maximal. One of those vertices must be reached after a finite amount of iterations, resulting in the loops termination.
I guess I will just comment "no same vertex can be selected as `current` twice", as that's the essential argument.

Reply:
Yes I'm ok with this.